File: CC/MakeACake_CodeUncode_2020.py
length = input()
str_arr = input().split(" ")
sort_arr= [int(i) for i in str_arr]
arr = sort_arr.copy()
printNewLine_flag = True

sort_arr.sort()
stack = []
mem = []

for j in sort_arr:
    stack.append(j)
    #print(j)

for j in arr:    
    if len(stack) != 0:
        #print("j: " + str(j) + " top: " + str(stack[-1]))
        if j < stack[-1]:
            mem.append(j)
            #print("j: " + str(j) + ", mem: " + str(mem))
            print()
        else:
            stack.pop()
            print(j, end="")
            if len(mem) != 0 and len(stack) != 0:
                mem.sort()
                while len(mem) != 0 and mem[-1]>=stack[-1]:
                    print(" " + str(mem.pop()), end="")
                    stack.pop()
                    printNewLine_flag = True
            if printNewLine_flag:
                print()
                printNewLine_flag = False
while len(stack) != 0:
    print(stack.pop())

File: Unkown/MakeACake_CodeUncode_2020.java
import java.io.*;
import java.util.*;

public class MakeACake_CodeUncode_2020 {

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int length = sc.nextInt();
        ArrayList<Integer> arr = new ArrayList<>();
        ArrayList<Integer> sort_arr = new ArrayList<>();
        ArrayList<Integer> mem = new ArrayList<>();
        Stack<Integer> stack = new Stack<>();
        Boolean flag = true;
        
        for(int i=0; i<length; i++){
            int x = sc.nextInt(); 
            arr.add(x);
            sort_arr.add(x);   
        }
        
        Collections.sort(sort_arr);
        
        for(int i=0; i<length; i++)
            stack.push(sort_arr.get(i));        
        
        for(int i=0; i<length; i++){
            if(!stack.empty()){
                if(arr.get(i) < stack.peek()){
                    mem.add(arr.get(i));
                    Collections.sort(mem);
                    System.out.println("");
                }
                else{
                    stack.pop();
                    System.out.print(arr.get(i));
                    if(mem.size() != 0 && !stack.empty()){
                        while(mem.size() != 0 && mem.get(mem.size()-1)>=stack.peek()){
                            System.out.print(" " + mem.get(mem.size()-1));
                            mem.remove(mem.size()-1);
                            stack.pop();
                            flag = false;
                        }
                    }
                    if (flag){
                        System.out.println();
                        flag = false;
                    }
                }
            }
        }
    }
}